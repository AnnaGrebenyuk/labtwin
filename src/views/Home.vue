<template>
  <div>
    <b-container>
      <b-row>
        <b-col sm md lg="6">
          <b-form-group label-cols="2" label="Sort by:">
            <b-form-select class="select-sort"
                      :value="sortOption"
                      :options="sortingOptions"
                      value-field="text"
                      @change="changeSortOption"/>
          </b-form-group>
        </b-col>
        <b-col sm md lg="6">
          <b-button class="btn-segment"
          v-for="s in segments"
          :key="s"
          @click="setSegment(s)"
          :variant="s === segment ? 'primary' : 'secondary'"
          >
            {{ s }}
          </b-button>
        </b-col>
      </b-row>
      <b-row align-h="center">
        <b-col lg="4" sm="4" md="4">
          <b-pagination-nav :link-gen="getPageLink" :number-of-pages="page.totalPages" use-router />
        </b-col>
      </b-row>
    </b-container>
    <events-list :events="events" :favoritesMode="false"/>
  </div>
</template>

<script>
import { mapState, mapActions, mapMutations } from 'vuex';
import {
  BContainer, BRow, BCol, BFormGroup, BFormSelect, BButton, BPaginationNav,
} from 'bootstrap-vue';
import EventsList from '../components/EventsList.vue';

export default {
  name: 'home',
  components: {
    EventsList, BContainer, BRow, BCol, BFormGroup, BFormSelect, BButton, BPaginationNav,
  },
  data: () => ({
    segments: ['Sports', 'Music', 'Arts & Theatre', 'All'],
    sortingOptions: [
      {
        text: 'Most Popular',
      },
      {
        text: 'Distance',
      },
      {
        text: 'Name: A - Z',
      },
      {
        text: 'Name: Z - A',
      },
      {
        text: 'Date',
      },
    ],
  }),
  computed: mapState(['events', 'page', 'sortOption', 'segment']),
  methods: {
    ...mapActions(['getEventsData', 'getGeolocation']),
    ...mapMutations(['addToFavorites', 'setSortOption', 'setSegment']),
    getPageLink(pageNumber) {
      return pageNumber === 1 ? '?' : `?page=${pageNumber}`;
    },
    changeSortOption(option) {
      this.setSortOption(option);
      this.getEventsData({
        pageNumber: this.$route.query.page,
        segment: this.selectedSegment,
        sortOption: option,
      });
    },
  },
  watch: {
    // eslint-disable-next-line func-names
    '$route.query.page': function (pageNumber) {
      this.getEventsData({
        pageNumber,
        segment: this.selectedSegment,
        sortOption: this.selectedSortOption,
      });
    },
    segment(segment) {
      this.getEventsData({ pageNumber: this.$route.query.page, segment });
    },
  },
  created() {
    this.getEventsData({
      pageNumber: this.$route.query.page,
      segment: this.selectedSegment,
      sortOption: this.selectedSortOption,
    });
  },
};
</script>
