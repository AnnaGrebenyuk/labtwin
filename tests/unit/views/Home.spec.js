import {
  createLocalVue,
  mount,
} from '@vue/test-utils';
import Vuex from 'vuex';
import Home from '@/views/Home.vue';

const localVue = createLocalVue();
localVue.use(Vuex);


describe('Home Page', () => {
  let store;
  let state;
  let wrapper;

  const getEventsData = jest.fn();
  const setSortOption = jest.fn();
  const setSegment = jest.fn();

  const $route = {
    query: {
      page: 1,
    },
  };
  beforeAll(() => {
    state = {
      page: {
      },
    };

    store = new Vuex.Store({
      state,
      actions: {
        getEventsData,
      },
      mutations: {
        setSortOption,
        setSegment,
      },
    });
  });

  it('renders correctly', () => {
    wrapper = mount(Home, {
      store,
      localVue,
      mocks: {
        $route,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('calls getEventsData on Sort option change', () => {
    wrapper.find('.select-sort').trigger('change', 'Distance');
    expect(getEventsData).toBeCalledTimes(1);
  });

  it('calls getEventsData on Segment selection', () => {
    wrapper.find('.btn-segment:nth-child(1)').trigger('click');
    expect(getEventsData).toBeCalledTimes(2);
  });

  it('calls getEventsData on route change', () => {
    $route.query.page = 2;
    expect(getEventsData).toBeCalledTimes(3);
  });

  it('generates correct links for different page numbers', () => {
    expect(wrapper.vm.getPageLink(1)).toEqual('?');
    expect(wrapper.vm.getPageLink(2)).toEqual('?page=2');
  });
});
