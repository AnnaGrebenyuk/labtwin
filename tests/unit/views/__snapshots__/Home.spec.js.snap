// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-sm col-md col-lg-6"
      >
        <fieldset
          class="form-group"
        >
          <div
            class="form-row"
          >
            <legend
              class="col-2 bv-no-focus-ring col-form-label"
              tabindex="-1"
            >
              Sort by:
            </legend>
            <div
              class="bv-no-focus-ring col"
              role="group"
              tabindex="-1"
            >
              <select
                class="select-sort custom-select"
              >
                <option
                  value="Most Popular"
                >
                  Most Popular
                </option>
                <option
                  value="Distance"
                >
                  Distance
                </option>
                <option
                  value="Name: A - Z"
                >
                  Name: A - Z
                </option>
                <option
                  value="Name: Z - A"
                >
                  Name: Z - A
                </option>
                <option
                  value="Date"
                >
                  Date
                </option>
              </select>
              <!---->
              <!---->
              <!---->
            </div>
          </div>
        </fieldset>
      </div>
       
      <div
        class="col-sm col-md col-lg-6"
      >
        <button
          class="btn btn-segment btn-secondary"
          type="button"
        >
          
          Sports
        
        </button>
        <button
          class="btn btn-segment btn-secondary"
          type="button"
        >
          
          Music
        
        </button>
        <button
          class="btn btn-segment btn-secondary"
          type="button"
        >
          
          Arts & Theatre
        
        </button>
        <button
          class="btn btn-segment btn-secondary"
          type="button"
        >
          
          All
        
        </button>
      </div>
    </div>
     
    <div
      class="row justify-content-center"
    >
      <div
        class="col-sm-4 col-md-4 col-lg-4"
      >
        <nav
          aria-hidden="false"
        >
          <ul
            aria-disabled="false"
            aria-label="Pagination"
            class="pagination b-pagination"
            role="menubar"
          >
            <li
              aria-hidden="true"
              class="page-item disabled"
              role="presentation"
            >
              <span
                aria-disabled="true"
                aria-label="Go to first page"
                class="page-link"
                role="menuitem"
              >
                «
              </span>
            </li>
            <li
              aria-hidden="true"
              class="page-item disabled"
              role="presentation"
            >
              <span
                aria-disabled="true"
                aria-label="Go to previous page"
                class="page-link"
                role="menuitem"
              >
                ‹
              </span>
            </li>
            <!---->
            <li
              class="page-item"
              role="presentation"
            >
              <a
                aria-checked="false"
                aria-label="Go to page 1"
                aria-posinset="1"
                aria-setsize="1"
                class="page-link"
                href="?"
                role="menuitemradio"
                tabindex="0"
              >
                1
              </a>
            </li>
            <!---->
            <li
              aria-hidden="true"
              class="page-item disabled"
              role="presentation"
            >
              <span
                aria-disabled="true"
                aria-label="Go to next page"
                class="page-link"
                role="menuitem"
              >
                ›
              </span>
            </li>
            <li
              aria-hidden="true"
              class="page-item disabled"
              role="presentation"
            >
              <span
                aria-disabled="true"
                aria-label="Go to last page"
                class="page-link"
                role="menuitem"
              >
                »
              </span>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
   
  <div
    class="container"
  >
    <div
      class="list-group"
    />
  </div>
</div>
`;
